drivers: udc: stm32: cleanup how PHY type and speed are determined

mathieuchopstm opened the pull request:
Rework how the STM32 USB driver determines the PHY type and operating speed. This approach is more solid, based on the `phys` property and `compatible` on the node itself, rather than looking at which nodes are enabled and hoping for the best. `PCD_PHY_xxx` and `PCD_SPEED_xxx` macros, which should be more portable, are also used now in place of the LL USB macros.

Commit 7ff59b5d855182b0a3e6c718f28caf39a64b0891 is cherry-picked from #89866; it is not *strictly* related to PR's contents, but HS does not work without it. Since this PR supersedes #89866, the commit could be merged as part of it; however, I don't mind #89866 being modified to only include this commit or a separate PR being opened...

This PR also partially supersedes #86100 (only a few cosmetic(?) changes are missing here).

bearsh committed:
drivers: udc_stm32: select Kconfig option UDC_DRIVER_HAS_HIGH_SPEED_SUPPORT

The driver supports high-speed capable controllers. Select
the appropriate Kconfig options to ensure that the stack does
not optimize away high-speed support.

Signed-off-by: Martin Gysel <me@bearsh.org>

mathieuchopstm commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -65,6 +65,78 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
+/**
+ * Returns the 'PCD_Speed' value for @p node_id
+ *
+ * For HS-capable instances, this may be one of:
+ *   - PCD_SPEED_HIGH: High-Speed mode
+ *   - PCD_SPEED_HIGH_IN_FULL: Full-Speed mode on HS instance
+ *
+ * PCD_SPEED_FULL is returned for instances that are not HS-capable.
```
Verified by running `testusb` on Nucleo-F207ZG (OTG_HS, embedded FS PHY)

erwango commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -41,27 +41,77 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
+ *
+ * The PHY DT node referenced by the 'phys' property of @p node_id
+ * is examined to determine the correct value:
+ *  - compatible 'st,stm32-usbphyc'    => embedded HS PHY
```
Wouldn't it clarify things to rename `st,stm32-usbphyc` to `st,stm32-usb-hs-phyc` ?

bearsh commented:
I hope I'll be able to test it later this week on my U5 based hardware...

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -41,27 +41,77 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
-#define UDC_STM32_FULL_SPEED             USB_OTG_SPEED_FULL
-#endif
+/** Evaluates to 1 if @p inst can run in High-Speed mode, 0 otherwise. */
+#define UDC_STM32_NODE_HS_CAPABLE(node_id)	DT_NODE_HAS_COMPAT(node_id, st_stm32_otghs)
```
This is written with multi-instancing in mind. The macro is (currently) used with `DT_DRV_INST(0)` as argument - the following at driver top will pick one of the two:
https://github.com/zephyrproject-rtos/zephyr/blob/2405ec5655e4c8c28ea407a280f14cd43061e53c/drivers/usb/udc/udc_stm32.c#L29-L38
(in practice, this means that if `otghs` and `otgfs` are present, the `otghs` takes priority)

When we get around to making the driver multi-instance capable, then we would evaluate this for each USB controller independently (or maybe we wouldn't need this macro anymore, depending on how that's implemented...)

erwango replied:
ok, right! For some reason my mind totally skipped the `node_id`

mathieuchopstm commented:
With cherry-picked https://github.com/zephyrproject-rtos/zephyr/pull/94001/commits/68b57ea8dca4cf3c0b6f62683d2610f841a3769f from #94001, I seem to get similar results as @nandojve : all `testusb` tests pass in FS mode, and all but test 13 pass in HS mode.

mathieuchopstm commented:
Rebased after merge of #95857

mathieuchopstm commented on dts/bindings/phy/st,stm32-embedded-usb-phy.yaml:
Really, this binding is only needed for the embedded PHY hardwired to `otghs` instances. Internal FS PHY is not always available for ST implementations of `otghs`; for example, STM32N6 or STM32U5 implementations of `otghs` are hardwired to an embedded HS PHY only (via UTMI+):
<img width="694" height="860" alt="image" src="https://github.com/user-attachments/assets/abccc204-6430-4d59-8f8f-7581b10f2fbb" />

(My understanding is that this should indeed have `FSPhyType = 2'b00` but I have not verified in HW)

On the other hand, some implementations of `otghs` in STM32 do have the FS PHY (and some straight up lack the ULPI interface: `OTG_HS2` on STM32H747). The FS PHYs don't require intervention from the application to work; the HS PHYs usually need at least one bit in `RCC` to be set to enable their clock, and sometimes more configuration (`st,stm32-usbphyc`/`st,stm32u5-otghs-phy`/... depending on which HS PHY was implemented)

Eventually, all PHY management should be moved out to a dedicated driver, and all existing PHYs would inherit from `st,stm32-embedded-usb`(?) - for now, special cases are grandfathered in to reduce the amount of changes.

mathieuchopstm replied:
After careful review, adding a new binding should be unnecessary: the embedded HS PHY on all supported STM32 already has a special compatible (`st,stm32u5-otghs-phy` for STM32U5, `st,stm32-usbphyc` for STM32N6 and _some_ STM32F7; STM32H7RS is not supported yet but would use `st,stm32u5-otghs-phy` as well).

I will rework this to drop the new compatible.

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1303,9 +1362,12 @@ static int udc_stm32_driver_init0(const struct device *dev)
-	if (cfg->speed_idx == 2) {
+
+#ifdef PCD_SPEED_HIGH
+	if (priv->pcd.Init.speed == PCD_SPEED_HIGH) {
```
Good remark - I did not notice the HAL updated this field. I will move this back to a dedicated field in the instance configuration.

mathieuchopstm replied:
Hmm. But I see that the DWC2 driver does:
https://github.com/zephyrproject-rtos/zephyr/blob/7b91e8f32c4e9c76c9e22efd979efeae131d5e5a/drivers/usb/udc/udc_dwc2.c#L1918-L1935

where `priv->enumspd` comes from `OTG_DSTS.ENUMSPD` (taking docs from STM32U5 RM):
```
Bits 2:1 ENUMSPD[1:0]: Enumerated speed
   Indicates the speed at which the OTG_HS controller has come up after speed detection through a chirp sequence.
   00: High Speed
   01: Full Speed
   11: Reserved
   Others: reserved
```

Since `pcd.Init.speed` is updated when enumeration finishes (in `HAL_PCD_IRQHandler()`), isn't that exactly the value we should be evaluating?

jfischer-no replied:
It should be evaluated in udc_stm32_device_speed() to get actual device speed. The udc_stm32_driver_init0() should not evaluate it, the controller is not operational at this point and should not be even recognizable by the host.

mathieuchopstm replied:
Nevermind my previous comment, I replied thinking it was applying to `udc_stm32_device_speed` and not `udc_stm32_driver_init0`... :sweat: 

I was thinking that `Init.speed` should be initialized because of this line:
https://github.com/zephyrproject-rtos/zephyr/blob/1a14623da6c0327e8937072f9de8b88a4ef42ac2/drivers/usb/udc/udc_stm32.c#L1097

But `priv_pcd_prepare` is called long after `udc_stm32_driver_init0()`, so indeed this is incorrect.

Will change driver to use a dedicated field in `config` as done before.

henrikbrixandersen commented:
Architecture WG, 2025-09-25:
- Consensus is to move forward with this PR first, @mathieuchopstm to address the review comments from @jfischer-no; @jfischer-no to respond to already posted questions from @mathieuchopstm
- After this PR is merged, the next step will be to have another look at https://github.com/zephyrproject-rtos/zephyr/pull/94001 from @nandojve
- Other PRs provided by the community attempting to solve the same issue(s) in the STM32 UDC shim driver shall be closed by ST with a thank you note and a reference to the PR(s) superseding them.
- Future plans for the STM32 UDC shim driver (reworking the compatibles to better describe the underlying controller, splitting the PHY code into a dedicated USB PHY driver, moving towards using the native DWC2 driver where possible) should be kept separate to this PR, ideally in a "meta" issue detailing/discussing these plans.

mathieuchopstm committed:
drivers: usb: udc: stm32: use HAL PCD macros for speed

Rework the STM32 USB UDC driver to use HAL PCD macros for speed selection,
as should have always been done since that is the library used by the shim
driver. Also dummy definitions to ensure cross-series compatibility
without many #if blocks in the driver.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
drivers: usb: udc: stm32: rework PWR and PHY initialization

Rework how the driver performs PWR configuration and initializes HS PHYs.

Originally, the initialization logic was split in half without any real
logic. There were also attempts to share some logic but this resulted in
code that was difficult to understand.

Initialize PWR and PHY in well separated steps, and use explicit series
and PHY type checks in each part to ensure the logic is easy to understand.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
drivers: usb: udc: stm32: enhance PHY type & operating speed determination

Rework how the STM32 UDC driver determines the PHY used by a given instance
and the speed at which the instance is able to and allowed to operate.

The PHY determination now uses the 'phys' property instead of looking at
whether nodes with a specific compatible were enabled. Similarly, the speed
determination takes into account the 'maximum-speed' property, along with
the selected PHY's and the instance's capabilities.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm commented:
FTR this has been tested to build, run and pass MSC tests on:
* `stm32h747i_disco//m7` (HS/FS) **+**
* `stm32n6570_dk` (HS/FS) **+**
* `nucleo_u5a5zj_q` (HS/FS)
* `b_u585i_iot02a` (FS only)
* `nucleo_l552ze_q` (FS only)

**+** indicates tests did not fully complete due to an error `LBA 0 is out of range` which I assume to be unrelated to the USB driver itself, but haven't investigated yet. I will do a pass of `testusb` later along with fix the small compliance issue...

mathieuchopstm commented:
@jfischer-no PTAL whenever possible :slightly_smiling_face:

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -53,23 +53,73 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
-#define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
-			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+/* Evaluates to 1 if PHY of 'usb_node' is an embedded HS PHY, 0 otherwise */
+#define UDC_STM32_PHY_HAS_EMBEDDED_HS_COMPAT(usb_node)	UTIL_OR(		\
+	DT_NODE_HAS_COMPAT(UDC_STM32_PHY(usb_node), st_stm32_usbphyc),		\
+	DT_NODE_HAS_COMPAT(UDC_STM32_PHY(usb_node), st_stm32u5_otghs_phy))
```
Maybe you want to use style from include/zephyr/devicetree.h
```c
#define UDC_STM32_PHY_HAS_EMBEDDED_HS_COMPAT(usb_node) \
	UTIL_OR(DT_NODE_HAS_COMPAT(UDC_STM32_PHY(usb_node), st_stm32_usbphyc), \
		DT_NODE_HAS_COMPAT(UDC_STM32_PHY(usb_node), st_stm32u5_otghs_phy))
```

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1081,13 +1081,19 @@ static int priv_clock_enable(void)
+		k_sleep(K_MSEC(100));
 	}
+#elif defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* Sequence to enable the power of the OTG HS on a stm32U5 serie : Enable VDDUSB */
```
I plan to clean this up when reworking this whole topic globally, so not necessary IMO.
